File: android/app/src/main/kotlin/com/example/kimoi/extractors/DoodExtractor.kt
package com.example.kimoi.extractors
import com.example.kimoi.model.Track
import com.example.kimoi.model.Video
import okhttp3.Headers
import java.net.HttpURLConnection
import java.net.URL

class DoodExtractor() {

    fun videoFromUrl(
        url: String,
        quality: String? = null,
        redirect: Boolean = true,
        externalSubs: List<Track> = emptyList(),
    ): Video? {
        val newQuality = quality ?: ("Doodstream" + if (redirect) " mirror" else "")

        return runCatching {
            val myURL = URL(url)
            val connection = myURL.openConnection() as HttpURLConnection
            connection.requestMethod = "GET"
            connection.setRequestProperty("Accept", "application/json")
            connection.setRequestProperty("User-agent", "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36")
            connection.connect()

            val inputStream = connection.inputStream
            val response = inputStream.bufferedReader().readText()

            val newUrl = if (redirect) connection.url.toString() else url

            val doodHost = Regex("https://(.*?)/").find(newUrl)!!.groupValues[1]
            val content = response
            if (!content.contains("'/pass_md5/")) return null
            val md5 = content.substringAfter("'/pass_md5/").substringBefore("',")
            val token = md5.substringAfterLast("/")
            val randomString = getRandomString()
            val expiry = System.currentTimeMillis()

            val videoUrlS = URL("https://$doodHost/pass_md5/$md5")
            val connect = videoUrlS.openConnection() as HttpURLConnection
            connect.requestMethod = "GET"
            connect.setRequestProperty("referer", newUrl)
            connect.connect()

            val impStream = connect.inputStream

            val videoUrlStart = impStream.bufferedReader().readText()

            val videoUrl = "$videoUrlStart$randomString?token=$token&expiry=$expiry"
            Video(newUrl, newQuality, videoUrl, server = "Doodstream" ,headers = doodHeaders(doodHost), subtitleTracks = externalSubs)
        }.getOrNull()
    }

    fun videosFromUrl(
        url: String,
        quality: String? = null,
        redirect: Boolean = true,
    ): List<Video> {
        val video = videoFromUrl(url, quality, redirect)
        return video?.let(::listOf) ?: emptyList<Video>()
    }

    private fun getRandomString(length: Int = 10): String {
        val allowedChars = ('A'..'Z') + ('a'..'z') + ('0'..'9')
        return (1..length)
            .map { allowedChars.random() }
            .joinToString("")
    }

    private fun doodHeaders(host: String) = Headers.Builder().apply {
        add("User-Agent", "Aniyomi")
        add("Referer", "https://$host/")
    }.build()
}

File: lib/src/utils/extractors/dood_extractor.dart
import 'dart:math';

import 'package:http/http.dart';
import 'package:kimoi/src/infrastructure/infrastructure.dart';
import 'package:kimoi/src/utils/extensions/extension.dart';

class DoodExtractor {
  final client = Client();

  Future<Video?> videoFromUrl(
    String url,
    String? quality, {
    bool redirect = true,
    List<Track> externalSubs = const [],
  }) async {
    final newQuality = quality ?? "Doodstream${redirect ? " mirror" : ""}";

    try {
      final response = await client.get(Uri.parse(url));

      // if (response.body.isNotEmpty && response.statusCode == 200)

      final newUrl = (redirect) ? response.request!.url.toString() : url;

      final doodHost = RegExp("https://(.*?)/").firstMatch(newUrl)!.group(1);
      final content = response.body.toString();
      if (!content.contains("'/pass_md5/")) return null;
      final md5 = content.substringAfter("'/pass_md5/").substringBefore("',");
      final token = md5.substringAfterLast("/");
      final randomString = getRandomString();
      final expiry = DateTime.now().millisecond;
      final videoUrlStart = await client.get(
        Uri.parse("https://$doodHost/pass_md5/$md5"),
        headers: {"referer": newUrl},
      );
      final videoUrl =
          "${videoUrlStart.body.toString()}$randomString?token=$token&expiry=$expiry";
      return Video(
          url: newUrl,
          quality: newQuality,
          videoUrl: videoUrl,
          headers: doodHeaders(doodHost!),
          subtitleTracks: externalSubs);
    } catch (e) {
      return null;
    }
  }

  Map<String, String> doodHeaders(String host) {
    Map<String, String> headers = {};
    headers
      ..addAll({"User-Agent": "Aniyomi"})
      ..addAll({"Referer": "https://$host/"});

    return headers;
  }

  String getRandomString({int length = 10}) {
    final random = Random();

    final allowedChars = [
      'A',
      'B',
      'C',
      'D',
      'E',
      'F',
      'G',
      'H',
      'I',
      'J',
      'K',
      'L',
      'M',
      'N',
      'O',
      'P',
      'Q',
      'R',
      'S',
      'T',
      'U',
      'V',
      'W',
      'X',
      'Y',
      'Z',
      'a',
      'b',
      'c',
      'd',
      'e',
      'f',
      'g',
      'h',
      'i',
      'j',
      'k',
      'l',
      'm',
      'n',
      'o',
      'p',
      'q',
      'r',
      's',
      't',
      'u',
      'v',
      'w',
      'x',
      'y',
      'z',
      '0',
      '1',
      '2',
      '3',
      '4',
      '5',
      '6',
      '7',
      '8',
      '9'
    ];

    final l = length <= 1 ? 1 : length;
    List<String> result = [];
    for (var i = 0; i < l; i++) {
      result.add(allowedChars[random.nextInt(allowedChars.length)]);
    }
    return result.join();
    // .joinToString("")
  }
}
